sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/perPartitionCircuitBreaker/GlobalPartitionEndpointManagerForPerPartitionCircuitBreaker.java

```diff
@@ -70,168 +77,192 @@ public void init() {
+                if (isFailureThresholdBreached.get()) {
 
-                isFailoverPossible.set(
-                    partitionLevelLocationUnavailabilityInfoAsVal.areLocationsAvailableForPartitionKeyRange(applicableRegionalRoutingContexts));
-            }
+                    UnmodifiableList<RegionalRoutingContext> applicableRegionalRoutingContexts = request.isReadOnlyRequest() ?
+                        this.globalEndpointManager.getApplicableReadRegionalRoutingContexts(request.requestContext.getExcludeRegions()) :
+                        this.globalEndpointManager.getApplicableWriteRegionalRoutingContexts(request.requestContext.getExcludeRegions());
 
-            request.requestContext.setPerPartitionCircuitBreakerInfoHolder(partitionLevelLocationUnavailabilityInfoAsVal.regionToLocationSpecificHealthContext);
-            return partitionLevelLocationUnavailabilityInfoAsVal;
-        });
+                    isFailoverPossible.set(
+                        partitionLevelLocationUnavailabilityInfoAsVal.areLocationsAvailableForPartitionKeyRange(applicableRegionalRoutingContexts));
+                }
 
-        // set to true if and only if failure threshold exceeded for the region
-        // and if failover is possible
-        // a failover is only possible when there are available regions left to fail over to
-        if (isFailoverPossible.get()) {
-            return;
-        }
+                request.requestContext.setPerPartitionCircuitBreakerInfoHolder(partitionLevelLocationUnavailabilityInfoAsVal.regionToLocationSpecificHealthContext);
+                return partitionLevelLocationUnavailabilityInfoAsVal;
+            });
 
-        if (logger.isWarnEnabled()) {
-            logger.warn("It is not possible to mark region {} as Unavailable for partition key range {}-{} and collection rid {} " +
-                    "as all regions will be Unavailable in that case, will remove health status tracking for this partition!",
-                this.globalEndpointManager.getRegionName(
-                    failedRegionalRoutingContext.getGatewayRegionalEndpoint(), request.isReadOnlyRequest() ? OperationType.Read : OperationType.Create),
-                resolvedPartitionKeyRangeForCircuitBreaker.getMinInclusive(),
-                resolvedPartitionKeyRangeForCircuitBreaker.getMaxExclusive(),
-                collectionResourceId);
-        }
+            // set to true if and only if failure threshold exceeded for the region
+            // and if failover is possible
+            // a failover is only possible when there are available regions left to fail over to
+            if (isFailoverPossible.get()) {
+                return;
+            }
+
+            if (logger.isWarnEnabled()) {
+                logger.warn("It is not possible to mark region {} as Unavailable for partition key range {}-{} and collection rid {} " +
```

Comment:
is this for partition key range or epk?
Also, shall we log the operation type here?